File: android/app/src/main/java/sol/vortexx/android/rooms/api/RoomsRepository.kt
package sol.vortexx.android.rooms.api

import kotlinx.coroutines.flow.Flow
import sol.vortexx.android.db.entities.RoomEntity

/** Feature contract for the rooms list / create / join. */
interface RoomsRepository {
    fun observe(): Flow<List<RoomEntity>>
    suspend fun refresh(): RefreshResult
    suspend fun create(name: String, isPrivate: Boolean): RoomResult
    suspend fun joinByInvite(inviteCode: String): RoomResult
    suspend fun leave(roomId: Long): Boolean
}

sealed interface RefreshResult {
    data class Ok(val count: Int) : RefreshResult
    data class Error(val reason: String) : RefreshResult
}

sealed interface RoomResult {
    data class Ok(val roomId: Long) : RoomResult
    data class Error(val code: String, val message: String) : RoomResult
}

File: ios/Modules/Sources/Rooms/api/RoomsRepository.swift
import Foundation
import DB

/// Rooms list / create / join feature contract.
public protocol RoomsRepository: Sendable {
    func observe() -> AsyncStream<[RoomRecord]>
    func refresh() async -> RefreshResult
    func create(name: String, isPrivate: Bool) async -> RoomResult
    func joinByInvite(_ code: String) async -> RoomResult
    func leave(_ roomId: Int64) async -> Bool
}

public enum RefreshResult: Sendable, Equatable {
    case ok(count: Int)
    case error(String)
}

public enum RoomResult: Sendable, Equatable {
    case ok(roomId: Int64)
    case error(code: String, message: String)
}
